http-server/src/main/java/org/triplea/server/access/AuthenticatedUser.java

```diff
@@ -22,11 +24,12 @@ public String getName() {
     throw new UnsupportedOperationException("Name lookup is not done on authentication.");
   }
 
-  public boolean isAdmin() {
-    return userRole.equals(UserRole.ADMIN);
-  }
-
-  public boolean isModerator() {
-    return userRole.equals(UserRole.MODERATOR);
+  public int getUserIdOrThrow() {
+    Preconditions.checkState(
+        !userRole.equals(UserRole.ANONYMOUS),
+        "Anonymous user is expected to have a null user id, "
+            + "it is the only case expected where user id would be null");
+    return Optional.ofNullable(userId)
+        .orElseThrow(() -> new AssertionError("Expected to have a non-null user id"));
```

Comment:
Not really understanding this method. It throws an error if the user role is anonymous or if the user has a null user id? Also maybe I'm just not used to seeing it written this way but seems a bit strange to create an optional of an Integer to then just immediately check if its null? Couldn't you just directly check if `userId == null`?

Also given this change, do any of the code paths that now call this method properly handle the expections or if this happens does it kill everything?

Reply:
> if the user role is anonymous or if the user has a null user id? 

Yes. We expect user id to be null only with anonymous accounts. If the user id were null, it's indirect to say "the user id was null", instead we can say "this user anonymous, they do have a user id"

To avoid predicting that we will always have the 4 roles, and in case we do not update this code, we check both situations. It can also be thought of as this way, we have a validity check, "user id != null", and the precondition check is documentation.

> Also maybe I'm just not used to seeing it written this way but seems a bit strange to create an optional of an Integer to then just immediately check if its null? Couldn't you just directly check if userId == null?

```
Optional.ofNullable(userId).orElseThrow(..)
```
Is the same as:
```
if(userId == null) {
  throw ...
} else {
 return userId;
}
```
I suppose the difference is imperative vs functional, one says what to do, the other says how to do it.

> Also given this change, do any of the code paths that now call this method properly handle the expections or if this happens does it kill everything?

Kills everything by design. The code paths where we could get a null user id would NPE out anyways. WE could put NPE guards everywhere, but in each case we want the code to terminate anyways. Having no user-id when we expect one is effectively a security breach.

Reply:
Let me know if that properly answers your questions @ron-murhammer

Reply:
Ok. I think my confusion is maybe the precondition message: "Anonymous user is expected to have a null user id, it is the only case expected where user id would be null". That reads kind of odd to me and not sure what its trying to tell me. If I saw that in say the logs after some sort of issue, I don't think I would understand what happened. Is there a better way to phrase that?

The other thing that strikes me as a little inconsistent is one of the 2 checks is done in a precondition while the other is then done as part of the return statement. Would it be more clear to check both as preconditions? Essentially more explicitly saying this method requires the user to not be anonymous and not have a null user ID?

Reply:
The precondition message should be improved, that is pretty clear.

Ideally we would never see these messages in logs. The idea is to enforce strong validity constraints and never let these error conditions get past testing, the software simply fails/blows up before we ever get to production. On the other side of the coin, it's 99% likely a developer bug/error if we ever have these bad states, so we should never really see these situations.

The precondition being documentation I think makes sense, it's better than a comment and more direct and adheres to 'fail-early' just slightly better. Checking both conditions as a precondition is a bit odd. The method checks for null and throws or returns a non-null ID, that is what the method does. Using a precondition to check that is a bit odd.

I'm avoiding the assumption a bit that the only case with a null user id is when the user is authenticated as anonymous.  We could have a bug somewhere that maybe leads to that situation. It's a bit the arrow of causality, an anonymous user implies a null user id, but a null user id might not always necessarily imply an anonymous user. Ideally it is bidirectional and basically assumes we have no bugs and are correct about enforcing this invariant. Hence the wiggle-room and why the two checks are not the same.